phoenix-core/src/main/java/org/apache/phoenix/index/GlobalIndexChecker.java

```diff
@@ -0,0 +1,426 @@
+            if (buildIndexScan == null) {
+                buildIndexScan = new Scan();
+                indexScan = new Scan(scan);
+                byte[] dataTableName = scan.getAttribute(PHYSICAL_DATA_TABLE_NAME);
+                byte[] indexTableName = region.getRegionInfo().getTable().getName();
+                dataHTable = ServerUtil.getHTableForCoprocessorScan(env,
+                        SchemaUtil.getPhysicalTableName(dataTableName, env.getConfiguration()));
+                if (indexMaintainer == null) {
+                    byte[] md = scan.getAttribute(PhoenixIndexCodec.INDEX_PROTO_MD);
+                    List<IndexMaintainer> maintainers = IndexMaintainer.deserialize(md, true);
+                    indexMaintainer = getIndexMaintainer(maintainers, indexTableName);
+                }
+                if (indexMaintainer == null) {
+                    throw new DoNotRetryIOException(
+                            "rebuildSingleIndexRow: IndexMaintainer is not included in scan attributes for " +
+                                    region.getRegionInfo().getTable().getNameAsString());
+                }
+                if (viewConstants == null) {
+                    viewConstants = IndexUtil.deserializeViewConstantsFromScan(scan);
+                }
+                buildIndexScan.setAttribute(BaseScannerRegionObserver.UNGROUPED_AGG, TRUE_BYTES);
+                buildIndexScan.setAttribute(PhoenixIndexCodec.INDEX_PROTO_MD, scan.getAttribute(PhoenixIndexCodec.INDEX_PROTO_MD));
+                buildIndexScan.setAttribute(BaseScannerRegionObserver.REBUILD_INDEXES, TRUE_BYTES);
+            }
+            // Rebuild the index row from the corresponding the row in the the data table
+            byte[] dataRowKey = indexMaintainer.buildDataRowKey(new ImmutableBytesWritable(indexRowKey), viewConstants);
+            buildIndexScan.withStartRow(dataRowKey, true);
+            buildIndexScan.withStopRow(dataRowKey, true);
+            buildIndexScan.setRaw(true);
+            ResultScanner resultScanner = dataHTable.getScanner(buildIndexScan);
+            if (resultScanner.next() == null) {
+                // The corresponding row does not exist in the data table. Need to delete the row from index if it is old enough
+                deleteRowIfAgedEnough(indexRowKey, ts);
+                row.clear();
+                return row;
+            }
+            // Scan the newly build row
+            indexScan.withStartRow(indexRowKey, true);
+            indexScan.withStopRow(indexRowKey, true);
+            RegionScanner regionScanner = region.getScanner(indexScan);
+            regionScanner.next(row);
+            regionScanner.close();
+            if (!row.isEmpty() && !verifyRowAndRemoveVerifiedCell(row)) {
+                deleteRowIfAgedEnough(indexRowKey, ts);
+                row.clear();
+            }
+            return row;
+        }
+
+        private boolean isVerifiedCell(Cell cell) {
```

Comment:
Does this function check for the empty cell length being 0 meaning nothing set yet? 
In that case should not this return false for length 0?
Recommend renaming it to isVerifiedSetCell

Reply:
Better idea: How about create a function that would return an Enum: EMPTY, SET_TRUE, SET_FALSE.
This way you can remove verifyRow and refactor verifyRowAndRemoveVerifiedCell and all use this function

Reply:
Cell length cannot be zero. 
verifyRow and verifyRowAndRemoveVerifiedCell serve different purposes. Verify row is defensive coding. If there are cases where the verified cell does not get added to the scan then verifyRow uses the HBase get operation to retrieve the verify cell